### PR TITLE
TIP-676: introduce a dedicated column for product identifier

### PR DESCRIPTION
**Pull request by jjanvier**
## Context
Currently, the identifier of the product is present in the values. It has some main advantage to let the identifier be enriched exactly as other attributes in the UI. But that's the only pro. Even PO don't think it should be treated as a regular attribute (see https://akeneo.slack.com/files/julien.janvier/F3PDZLMLK/Functional_impacts_on_single_storage).

But this has a major drawback from a technical point of view. We can't directly query a product by its identifier.

To solve this problem, we propose to add an identifier column in the product table. To avoid to break any functional behavior, an identifier value will also be added to the values during their loading. This value will be removed before the save of the product to avoid to duplicate the information.

We will also keep the possibility to query products by their attribute identifier. For instance, the user will still be able look for products with the `sku` "toto". But this search will redirect on the `identifier` property.

## Expected
- [x] the `identifier` column is added
- [x] the `findByIdentifer` methods now uses this column
- [x] the identifier is treated as a regular field by the `ProductQueryBuilder`
- [x] the value is added during the load of the product
- [x] the value is removed before the save of the product

## Notes

A temporary `identifier` PQB filter has been added. It will be removed when we'll implement the ES search feature. Its purpose is only to make the Behat and integration tests pass.

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Y
| Added Behats                      | Y
| Added integration tests           | Y
| Changelog updated                 | Y
| Review and 2 GTM                  | :clock1:
| Micro Demo to the PO (Story only) |- 
| Migration script                  | will be done later
| Tech Doc                          | will be done later

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductRepository.php [hunk 1]
```diff
@@ -389,23 +389,7 @@ function ($row) {
...
-        $qb->where($rootAlias . '.rawValues LIKE :identifierCriteria');
-        $qb->setParameter('identifierCriteria', '%' . $identifier . '%');
-        $result = $qb->getQuery()->execute();
-
-        if (empty($result)) {
-            return null;
-        }
-
-        return reset($result);
+        return $this->findOneBy(['identifier' => $identifier]);
```
Comment (damien-carcel) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductRepository.php:392]: Oh yeah!

**Review by SamirBoulil — APPROVED** [review]
GTM

**Review by damien-carcel — APPROVED** [review]

File: features/filter/identifier.feature [hunk 2]
```diff
@@ -23,3 +23,25 @@ Feature: Filter on identifier
...
+      | sku     |
+      | BOOTBXS |
+      | BOOTWXS |
+      | BOOTBS  |
+      | BOOTBL  |
+      | MUGRXS  |
+    Then I should get the following results for the given filters:
+      | filter                                                                                                                                    | result                                     |
+      | [{"field":"identifier", "operator":"STARTS WITH",      "value": "BOOT"   }]                                                               | ["BOOTBXS", "BOOTWXS", "BOOTBS", "BOOTBL"] |
+      | [{"field":"identifier", "operator":"STARTS WITH",      "value": "BOOT"   }]                                                               | ["BOOTBXS", "BOOTWXS", "BOOTBS", "BOOTBL"] |
```
Comment (skeleton) [inline features/filter/identifier.feature]: duplication

File: src/Pim/Component/Catalog/Model/ProductInterface.php [hunk 3]
```diff
@@ -44,13 +44,19 @@ public function setId($id);
...
      *
-     * @throws MissingIdentifierException if no identifier could be found
-     *
-     * @return ProductValueInterface the identifier of the product
+     * @return string
      */
     public function getIdentifier();
 
     /**
+     * @param $identifier
```
Comment (skeleton) [inline src/Pim/Component/Catalog/Model/ProductInterface.php]: missing type

File: src/Pim/Component/Catalog/Model/ProductInterface.php [hunk 4]
```diff
@@ -44,13 +44,19 @@ public function setId($id);
...
+     *
+     * @return ProductValueInterface
```
Comment (skeleton) [inline src/Pim/Component/Catalog/Model/ProductInterface.php]: nope, this method returns nothing

**Review by skeleton — APPROVED** [review]

File: features/filter/identifier.feature (hunk 2 above)
Comment (jjanvier) [reply features/filter/identifier.feature]: done

File: src/Pim/Component/Catalog/Model/ProductInterface.php (hunk 3 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Model/ProductInterface.php]: done

File: src/Pim/Component/Catalog/Model/ProductInterface.php (hunk 4 above)
Comment (jjanvier) [reply src/Pim/Component/Catalog/Model/ProductInterface.php]: fixed the method

**Pull request merged**